profiler/profile.go

```diff
@@ -113,42 +206,80 @@ func (b *batch) addProfile(p *profile) {
-	switch t {
-	case HeapProfile:
-		return heapProfile(p.cfg)
-	case CPUProfile:
-		return p.cpuProfile()
-	case MutexProfile:
-		return mutexProfile(p.cfg)
-	case BlockProfile:
-		return blockProfile(p.cfg)
-	case GoroutineProfile:
-		return goroutineProfile(p.cfg)
-	case expGoroutineWaitProfile:
-		return goroutineWaitProfile(p.cfg)
-	case MetricsProfile:
-		return p.collectMetrics()
-	default:
-		return nil, errors.New("profile type not implemented")
-	}
-}
-
-// writeHeapProfile writes the heap profile; replaced in tests
-var writeHeapProfile = pprof.WriteHeapProfile
-
-func heapProfile(cfg *config) (*profile, error) {
-	var buf bytes.Buffer
+func (p *profiler) runProfile(pt ProfileType) ([]*profile, error) {
 	start := now()
-	if err := writeHeapProfile(&buf); err != nil {
+	t := pt.lookup()
+	// Collect the original profile as-is.
+	data, err := t.Collect(t, p)
+	if err != nil {
 		return nil, err
 	}
+	profs := []*profile{{
+		name: t.Filename,
+		data: data,
+	}}
+	// Compute the deltaProf (will be nil if not enabled for this profile type).
+	deltaStart := time.Now()
+	deltaProf, err := p.deltaProfile(t, data)
+	if err != nil {
+		return nil, fmt.Errorf("delta profile error: %s", err)
+	}
+	// Report metrics and append deltaProf if not nil.
 	end := now()
-	tags := append(cfg.tags, HeapProfile.Tag())
-	cfg.statsd.Timing("datadog.profiler.go.collect_time", end.Sub(start), tags, 1)
+	tags := append(p.cfg.tags, pt.Tag())
+	// TODO(fg) stop uploading non-delta profiles in the next version of
+	// dd-trace-go after delta profiles are released.
+	if deltaProf != nil {
+		profs = append(profs, deltaProf)
+		p.cfg.statsd.Timing("datadog.profiler.go.delta_time", end.Sub(deltaStart), tags, 1)
+	}
+	p.cfg.statsd.Timing("datadog.profiler.go.collect_time", end.Sub(start), tags, 1)
+	return profs, nil
+}
+
+// deltaProfile derives the delta profile between curData and the previous
+// profile. For profile types that don't have delta profiling enabled, it
+// simply returns nil, nil.
+func (p *profiler) deltaProfile(t profileType, curData []byte) (*profile, error) {
+	// Not all profile types use delta profiling, return nil if this one doesn't.
+	if t.Delta == nil {
+		return nil, nil
+	}
+	curProf, err := pprofile.ParseData(curData)
+	if err != nil {
+		return nil, fmt.Errorf("delta prof parse: %v", err)
+	}
+	var deltaData []byte
+	if prevProf := p.prev[t.Type]; prevProf == nil {
+		// First time deltaProfile gets called for a type, there is no prevProf. In
+		// this case we emit the current profile as a delta profile.
+		deltaData = curData
+	} else {
+		// Delta profiling is also implemented in the Go core, see commit below.
+		// Unfortunately the core implementation isn't resuable via a API, so we do
+		// our own delta calculation below.
+		// https://github.com/golang/go/commit/2ff1e3ebf5de77325c0e96a6c2a229656fc7be50#diff-94594f8f13448da956b02997e50ca5a156b65085993e23bbfdda222da6508258R303-R304
```

Comment:
I didn't ask the Go project, but I assume that they are not interested because the google cloud profiler is also using its own [delta profile implementation](https://github.com/googleapis/google-cloud-go/blob/master/profiler/heap.go#L51-L77).